### PR TITLE
Migrate examples to notebooks 

### LINKED ISSUE #228: Migrate current examples to notebooks

### PR DESCRIPTION
**Pull request by waltsims**
closes #228

### CODE REVIEW

**Comment by waltsims** [comment]
- [ ] update simulated data file on Google Drive and respective hash in code.

**Comment by waltsims** [comment]
- [ ] update Readme example section

**Comment by waltsims** [comment]
This closes #192 by removing custom reconstruction code.

**Comment by faridyagubbayli** [comment]
@waltsims here is the new hash: `1lGFTifpOrzBYT4Bl_ccLu_Kx0IDxM0Lv` (corresponds to the updated sensor data file [here](https://drive.google.com/file/d/1lGFTifpOrzBYT4Bl_ccLu_Kx0IDxM0Lv/view?usp=sharing)).

**Comment by waltsims** [comment]
- [ ] Move beamforming example into subdirectory
- [ ] Final clean up of examples

**Comment by waltsims** [comment]
- [ ] fix windows example test timeout

**Comment by waltsims** [comment]
The us_bmode_linear_transducer is missing the harmonic filtering plots since they are not implemented in the filtering method.

**Comment by waltsims** [comment]
Question: Should we enforce that we clear all notebook output data to keep our git graph clear of binary data? Currently cell outputs are quite large.

**Comment by waltsims** [comment]
- [ ] update main readme figure

**Comment by waltsims** [comment]
Removes uff and therefore closes #36

File: examples/us_bmode_linear_transducer/us_bmode_linear_transducer.ipynb [hunk 1]
```diff
@@ -0,0 +1,451 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/us_bmode_linear_transducer/us_bmode_linear_transducer.ipynb:8]: As the message says you can get around this by restarting the session and running again I believe.

File: examples/at_array_as_source/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Defining A Source Using An Array Transducer Example
+
+This example provides a demonstration of using the kWaveArray class to define an array transducer with three arc-shaped elements without staircasing errors.
+
+For a more detailed discussion of this example and the underlying techniques, see E. S. Wise, B. T. Cox, J. Jaros, & B. E. Treeby (2019). Representing arbitrary acoustic source and sensor distributions in Fourier collocation methods. The Journal of the Acoustical Society of America, 146(1), 278-288. https://doi.org/10.1121/1.5116132.
+
+To read more, visit the [original example page](www.k-wave.org/documentation/example_at_array_as_source.php).
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_source/README.md:7]: This works for me. I don't believe the website is secure. Can you navigate to that page on the k-Wave webpage?

File: examples/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Examples
+
+Many examples from k-wave-python are Python mirrors of the [large collection of great examples](http://www.k-wave.org/documentation/k-wave_examples.php) from the original k-wave project.
+All examples are included as python files or notebooks in a subdirectory of the example directory.
+Every example has a short readme.md file which briefly describes the purpose of the example.
+
+## List of Examples
+
+- [Array as a sensor](at_array_as_sensor/) ([original example](http://www.k-wave.org/documentation/example_at_array_as_sensor.php), [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/waltsims/k-wave-python/blob/migrate-examples-to-notebooks/examples/at_array_as_sensor/at_array_as_sensor.ipynb))
+- [Array as a source](array_as_source/) ([original example](http://www.k-wave.org/documentation/example_at_array_as_source.php), [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/waltsims/k-wave-python/blob/migrate-examples-to-notebooks/examples/at_array_as_source/at_array_as_source.ipynb))
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/README.md:10]: For me this works on the current head of the branch in the web browser. Have you pulled the most current changes?

File: examples/at_array_as_sensor/at_array_as_sensor.ipynb [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    "    karray.add_arc_element(element_pos[:, idx], radius, diameter, focus_pos)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb]: done

File: examples/at_array_as_sensor/at_array_as_sensor.ipynb [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    "# make a small disc in the top left of the domain\n",
+    "source.p0 = make_disc(N, Vector([N.x / 4 + x_offset, N.y / 4]), 4)\n",
+    "source.p0[99:119, 59:199] = 1\n",
+    "logical_p0 = source.p0.astype(bool)\n",
+    "sensor = kSensor()\n",
+    "sensor.mask = element_pos"
+   ]
+  },
+  {
+   "cell_type": "code",
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb:93]: That seems to be coming from [here](https://github.com/waltsims/k-wave-python/blame/a80512851ef562ea90a1050bf962e9bda08a2fce/kwave/utils/io.py#L254C18-L254C18). I don't remember adding this warning. @faridyagubbayli Do you remember this?

I think this is unrelated to this PR, and if we want to address it we should open a new issue.

File: examples/at_array_as_sensor/at_array_as_sensor.ipynb [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    "colors = [\n",
+    "    (1, 1, 1),  # White (Background)\n",
+    "    (233/255, 131/255, 0/255),  # Orange (Sensor)\n",
+    "    (254/255, 221/255, 92/255),  # Yellow (Sources)\n",
+    "    (0.8, 0.8, 0.8),  # Light Grey (PML Mask)\n",
+    "]\n",
+    "cmap = plt.matplotlib.colors.ListedColormap(colors)\n",
+    "\n",
+    "fig, ax = plt.subplots()\n",
+    "c = ax.pcolormesh(combined_mask, cmap=cmap, shading='auto')\n",
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb:169]: I attempted to clean this up. Let me know what you think.

File: examples/at_array_as_source/at_array_as_source.ipynb [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    "plt.ylabel('y-position [m]')\n",
+    "plt.title('Source and PML Masks')\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_source/at_array_as_source.ipynb]: I removed it.

File: examples/at_linear_array_transducer/at_linear_array_transducer.ipynb [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    "time_delays = -(np.sqrt((ids * element_pitch) ** 2 + source_focus ** 2) - source_focus) / c0\n",
+    "time_delays = time_delays - min(time_delays)\n",
+    "\n",
+    "source_sig = source_amp * tone_burst(1 / kgrid.dt, source_f0, source_cycles,\n",
+    "                                      signal_offset=np.round(time_delays / kgrid.dt).astype(int))\n",
+    "karray = kWaveArray(bli_tolerance=0.05, upsampling_rate=10)\n",
+    "\n",
+    "for ind in range(element_num):\n",
+    "    x_pos = 0 - (element_num * element_pitch / 2 - element_pitch / 2) + ind * element_pitch\n",
+    "    karray.add_rect_element([x_pos, 0, kgrid.z_vec[0]], element_width, element_length, rotation)\n",
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_linear_array_transducer/at_linear_array_transducer.ipynb:97]: I have already opened #231 to address this.

File: examples/at_linear_array_transducer/at_linear_array_transducer.ipynb [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    "grid_size_x = 40e-3\n",
+    "grid_size_y = 20e-3\n",
+    "grid_size_z = 40e-3\n",
+    "ppw = 3\n",
+    "t_end = 35e-6\n",
+    "cfl = 0.5\n"
+   ]
+  },
+  {
+   "cell_type": "code",
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_linear_array_transducer/at_linear_array_transducer.ipynb:62]: yes! with %%capture. I've done this.

File: examples/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/README.md]: fixed it.

**Comment by waltsims** [comment]
- [ ] update branch name in colab links to master before merging

File: examples/at_array_as_sensor/at_array_as_sensor.ipynb [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install k-wave-python &> /dev/null"
+   ]
+  },
+  {
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb:13]: Me too. I'll open an issue.

File: examples/at_array_as_sensor/at_array_as_sensor.ipynb [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+   "outputs": [],
+   "source": [
+    "simulation_options = SimulationOptions(\n",
+    "    save_to_disk=True,\n",
+    "    data_cast='single',\n",
+    ")\n",
+    "\n",
+    "execution_options = SimulationExecutionOptions(is_gpu_simulation=True)\n",
+    "output = kspaceFirstOrder2D(kgrid, source, sensor, medium, simulation_options, execution_options)\n",
+    "_, _, reorder_index = cart2grid(kgrid, element_pos)\n",
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb:108]: I added a comment

File: examples/at_array_as_source/at_array_as_source.ipynb [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    "# Create the animation\n",
+    "ani = FuncAnimation(fig, update, frames=kgrid.Nt, interval=5, blit=False, repeat=True)"
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_source/at_array_as_source.ipynb:200]: I updated the max size in rcparams

File: examples/at_array_as_sensor/at_array_as_sensor.ipynb [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+    "pml_mask[:, :pml_size] = 1\n",
+    "pml_mask[-pml_size:, :] = 1\n",
+    "pml_mask[:, -pml_size:] = 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Visualization"
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb:132]: They did not in the matlab example. I've added it, but I personally don't think it is very expressive.

File: examples/at_array_as_sensor/at_array_as_sensor.ipynb [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+    "    (0.8, 0.8, 0.8),  # Light Grey (PML Mask)\n",
+    "]\n",
+    "cmap = plt.matplotlib.colors.ListedColormap(colors)\n",
+    "\n",
+    "fig, ax = plt.subplots()\n",
+    "c = ax.pcolormesh(combined_mask, cmap=cmap, shading='auto')\n",
+    "plt.axis('image')\n",
+    "\n",
+    "# Define labels for the colorbar\n",
+    "labels = {\n",
```
[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb:177]: I'm not sure what you mean by arcs here?

[the remark this answers is not in the document]
Comment (waltsims) [reply examples/at_array_as_sensor/at_array_as_sensor.ipynb:179]: So like the circle through all the  transducer elements?

**Pull request merged**